File: ch02-datastructures-in-c/02-linked-list/0004-deletion-of-node-from-linked-list/deletion-of-node-from-linked-list-solution-01.c
/*
** Deletes a node from a linked list
*/

// Best Case Time Complexity = O(1) => when node to be deleted is head node
// Worst Case Time Complexity = O(N) => when node to be deleted is last node
// Space Complexity = O(N)

#include <stdio.h>
#include <stdlib.h>

struct node {
    int data;
    struct node *next;
};

void deleteNode(struct node **head, int key) {
    struct node *temp;
    if ((*head)->data == key) {
        temp = *head;
        *head = (*head)->next;
        free(temp);
    }
    else {
        struct node *current = *head;
        while (current->next != NULL) {
            if (current->next->data == key) {
                temp = current->next;
                current->next = current->next->next;
                free(temp);
                break;
            }
            else {
                current = current->next;
            }
        }
    }
}

void addLast(struct node **head, int val) {
    struct node *newNode = malloc(sizeof(struct node));
    newNode->data = val;
    newNode->next = NULL;
    
    if (*head == NULL)
        *head = newNode;
    else {
        struct node *lastNode = *head;
        while (lastNode->next != NULL)
            lastNode = lastNode->next;
        lastNode->next = newNode;
    }
}

void printList(struct node *head) {
    struct node *temp = head;
    printf("Traversing the linked list to print data of each node...\n");
    while (temp != NULL) {
        printf("%d ", temp->data);
        temp = temp->next;
    }
}

int main()
{
    struct node *head = NULL;
    
    addLast(&head, 10);
    addLast(&head, 20);
    addLast(&head, 30);
    
    int key;
    printf("Enter data of Node to be deleted: ");
    scanf("%d", &key);
    
    deleteNode(&head, key);
    
    printList(head);

    return 0;
}

File: ch03-datastructures-in-python/02-singly-linked-list/0004-deletion-of-node-from-linked-list/deletion-of-node-from-linked-list-solution-01.py
#!/usr/bin/env python3

# Searches a node in a linked list

# Best Case Time Complexity = O(1) => when node to be deleted is head node
# Worst Case Time Complexity = O(N) => when node to be deleted is last node
# Space Complexity = O(1)


class Node:
    def __init__(self, data):
        self.data = data
        self.next = None


class LinkedList:
    def __init__(self):
        self.head = None


def printList(self):
    print("Traversing the linked list to print data of each node...")
    node = self.head

    while node:
        print(node.data)
        node = node.next


def addLast(self, val):
    newNode = Node(val)

    # get last node
    if self.head == None:
        self.head = newNode
    else:
        lastNode = self.head

        while lastNode.next:
            lastNode = lastNode.next

        # add newNode to last node
        lastNode.next = newNode


def delete(self, key):
    if self.head != None:
        if self.head.data == key:
            temp = self.head
            self.head = self.head.next
            del temp
        else:
            current_node = self.head
            while current_node.next:
                if current_node.next.data == key:
                    temp = current_node.next
                    current_node.next = current_node.next.next
                    del temp
                    break
                else:
                    current_node = current_node.next
            return
    else:
        return


if __name__ == '__main__':

    llist = LinkedList()

    # add nodes at the beginning of linked list
    addLast(llist, 10)
    addLast(llist, 20)
    addLast(llist, 30)
    addLast(llist, 40)
    printList(llist)

    key = int(input("Enter data of Node to be deleted: "))
    delete(llist, key)

    printList(llist)
